fdbserver/BlobMigrator.actor.cpp

```diff
@@ -30,43 +30,273 @@
 #include "fdbclient/KeyRangeMap.h"
 #include "fdbclient/SystemData.h"
 #include "fdbclient/NativeAPI.actor.h"
+#include "fdbclient/ManagementAPI.actor.h"
 #include "fdbserver/ServerDBInfo.actor.h"
 #include "fdbserver/WaitFailure.h"
-
+#include "fdbserver/MoveKeys.actor.h"
+#include "fdbserver/BlobGranuleServerCommon.actor.h"
 #include "flow/actorcompiler.h" // has to be last include
+#include "flow/network.h"
+#include <algorithm>
+#include <string>
+
+#define ENABLE_DEBUG_MG true
+
+template <typename... T>
+static inline void dprint(fmt::format_string<T...> fmt, T&&... args) {
+	if (ENABLE_DEBUG_MG)
+		fmt::print(fmt, std::forward<T>(args)...);
+}
 
 // BlobMigrator manages data migration from blob storage to storage server. It implements a minimal set of
 // StorageServerInterface APIs which are needed for DataDistributor to start data migration.
 class BlobMigrator : public NonCopyable, public ReferenceCounted<BlobMigrator> {
 public:
 	BlobMigrator(Reference<AsyncVar<ServerDBInfo> const> dbInfo, BlobMigratorInterface interf)
-	  : blobMigratorInterf(interf), actors(false) {
+	  : interf(interf), actors(false) {
 		if (!blobConn.isValid() && SERVER_KNOBS->BG_METADATA_SOURCE != "tenant") {
 			blobConn = BlobConnectionProvider::newBlobConnectionProvider(SERVER_KNOBS->BG_URL);
 		}
 		db = openDBOnServer(dbInfo, TaskPriority::DefaultEndpoint, LockAware::True);
 	}
 	~BlobMigrator() {}
 
+	// Start blob migration
 	ACTOR static Future<Void> start(Reference<BlobMigrator> self) {
-		self->actors.add(waitFailureServer(self->blobMigratorInterf.waitFailure.getFuture()));
+		if (!isFullRestoreMode()) {
+			return Void();
+		}
+		wait(delay(10)); // TODO need to wait for a signal for readiness of blob manager
```

Comment:
What is the specific race here? That the migrator doesn't know what granules to read from until the manager is recovered?

Reply:
Yeah. The migrator need to wait for the blob manager to finish the recover. I plan to add a flag to system key space for the migrator to wait until it's done.

Reply:
Gotcha.
That would work, but it might actually be easier with an rpc request, since the blob manager already has some internal parts that wait on recovery and that it can signal when it is done. Up to you though.

Reply:
what happens if its not ready after 10 seconds? will it retry after another 10 seconds?

Reply:
Currently it simply fails if blob manager is not ready after 10 seconds.. My next PR will address this problem. It should wait and retry until blob manager is ready.. Just didn't want to make this PR too big.